File: Homework 1/Codes/22101001_ahmed_ghulam.py
#! /usr/bin/env python

def boolean_demo():
    # Boolean operators
    print("Boolean Operators:")
    print("True and False:", True and False)
    print("True or False:", True or False)
    print("not True:", not True)

    # Data types and Boolean values
    # Shows How different types (integers, lists, dictionaries) behave in boolean contexts.
    print("\nData Types and Boolean Values:")
    print("1 (Truthy) and 0 (Falsy):", 1 and 0)
    print("[] (Falsy) or {} (Falsy):", [] or {})
    print("[1, 2, 3] (Truthy) or {} (Falsy):", [1, 2, 3] or {})

    # Operator precedence
    # Shows how and takes precedence over or.
    print("\nOperator Precedence:")
    print("False or True and False:", False or True and False)  # Equivalent to False or (True and False)

    # Associativity
    # Shows how operators of the same precedence are evaluated left to right.
    print("\nAssociativity:")
    print("False and False or True and True:", False and False or True and True)  # Equivalent to ((False and False) or True) and True

    # Evaluation order
    # Shows the left-to-right evaluation order, particularly in the context of short-circuiting.
    print("\nEvaluation Order:")
    print("False and (print('Hello') or True):", False and (print('Hello') or True))  # 'Hello' is not printed

    # Short-circuit evaluation
    # Shows how Python doesn't evaluate the second operand of and/or if the first operand is sufficient to determine the result.
    print("\nShort-Circuit Evaluation:")
    print("True or (print('World') and False):", True or (print('World') and False))  # 'World' is not printed

boolean_demo()

File: Homework 1/Codes/22101001_ahmed_ghulam.lua
function booleanDemo()
    -- Boolean operators
    print("Boolean Operators:")
    print("true and false:", true and false) -- AND
    print("true or false:", true or false)      -- OR
    print("not true:", not true)                   -- NOT

    -- Data types and Boolean values
    -- This part demonstrates how Lua handles different values in boolean contexts.
    print("\nData Types and Boolean Values:")
    print("1 (Truthy) and 0 (Falsy):", 1 and 0) -- In Lua, all values except false and nil are truthy
    print("'string' (Truthy) or '' (Falsy):", 'string' or '') -- Strings are always truthy

    -- Operator precedence
    -- Shows how and takes precedence over or
    print("\nOperator Precedence:")
    print("false or true and false:", false or true and false) -- Equivalent to false or (true and false)

    -- Associativity
    -- Shows that and & or are left-associative.
    print("\nAssociativity:")
    print("true and false or true and true:", true and false or true and true) -- Equivalent to ((true and false) or true) and true

    -- Evaluation order
    -- Shows left-to-right evaluation, with the sideEffect() function not being called due to short-circuiting.
    print("\nEvaluation Order:")
    print("false and sideEffect() or true:", false and sideEffect() or true) -- sideEffect is not called due to short-circuit

    -- Short-circuit evaluation
    -- Shows how Lua does not evaluate the second operand of and or or if the first operand is sufficient to determine the result.

    print("\nShort-Circuit Evaluation:")
    print("true or sideEffect() and false:", true or sideEffect() and false) -- sideEffect is not called due to short-circuit
end

function sideEffect()
    print("Side effect function called")
    return true
end

booleanDemo()